### PR TITLE
FLOW-1619: Fix EMPTY operator for Objects and List iteration

### PR DESCRIPTION
**Pull request by ianhaycox**
Implemented IS_EMPTY for ContentObject

Re-factored getStateValue to move the GET_FIRST and GET_NEXT list iteration operations to Operations.ts. Also corrected an issue where for an Object property where getStateValue did not honour typeElementPropertyId and returned the whole object rather than the property.

Some fixes for the internaId/ExternalId problem when Offline and no externalId exists. This is to ensure the REMOVE functionality removes the correct item when Offline.

The ADD functionality is still broken when testing for duplicate inserts. This is again the internal/external ID problem. Working on that still.

### CODE REVIEW

File: js/services/Rules.ts [hunk 1]
```diff
@@ -69,13 +69,16 @@ const Rules = {
         let result = false;
 
         for (const rule of rules) {
-            const contentType = manywho.component.contentTypes.string;
 
             let left = snapshot.getValue(rule.leftValueElementToReferenceId);
-            left = getStateValue(rule.leftValueElementToReferenceId, left.typeElementId, left.contentType, '') || left;
+            // TODO - What to do if the operands have different contentTypes ? E.g.
```
Comment (jamesbrattBoomi) [inline js/services/Rules.ts:74]: I'm not 100% sure on this one. I think when trying to compare objects/lists with primitives, then an exception should probably be thrown. I'm not sure whether the expected behaviour is to use type coercion on comparing primitives or whether we support a deep diff on comparing objects.

File: js/models/State.ts [hunk 2]
```diff
@@ -28,39 +28,17 @@ export const StateInit = (state: IState) => {
 
 /**
  * @param id
- * @param typeElementId
- * @param contentType
- * @param command
  */
-export const getStateValue = (id: any, typeElementId: string, contentType: any, command: string) => {
+export const getStateValue = (id: any) => {
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/models/State.ts]: Done. Also added IFlow to some others

File: js/models/State.ts [hunk 3]
```diff
@@ -28,39 +28,17 @@ export const StateInit = (state: IState) => {
...
-                contentType, manywho.component.contentTypes.object, true) && value.objectData && value.objectData.length > 1) {
-                // TODO: log error;
-            }
-
-            if (id.typeElementPropertyId && value.objectData && value.objectData.length > 0) {
-                const property = value.objectData[0].properties.find(prop => prop.typeElementPropertyId === id.typeElementPropertyId);
-                if (property) {
-                    value.contentValue = property.contentValue ? property.contentValue : null;
-                    value.objectData = property.objectData ? property.objectData : null;
-                    value.index = 0;
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/models/State.ts:54]: The index isn't removed, it is used as the 'current pointer' to the array we are iterating with GET_NEXT, which is why it is saved in state so we can keep track of the loop.

File: js/services/Operation.ts [hunk 4]
```diff
@@ -83,15 +83,27 @@ export const executeOperation = (operation: any, state: IState, snapshot: any) =
...
-            );
+            const valueToReferenceStateValue = getStateValue(operation.valueElementToReferenceId);
             if (valueToReferenceStateValue) {
                 valueToReference = valueToReferenceStateValue;
             }
+
+            if (valueToReference.objectData) {
+                // Handle list iteration noting that GET_NEXT will get the first element if GET_FIRST has not been called beforehand
+                // Note that for iteration we need to persist the current array index by calling setStateValue()
+                if (manywho.utils.isEqual(operation.valueElementToReferenceId.command, 'GET_FIRST', true)) {
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/services/Operation.ts]: As discussed in Slack - it wasn't really any benefit to readability, so left as is.

File: js/services/Rules.ts (hunk 1 above)
Comment (ianhaycox) [reply js/services/Rules.ts:94]: Going to leave this one for now and tackle it later if it becomes an issue.  There's a lot of refactoring and error checking work to do, let alone work out the semantics of type conversions.

**Comment by ianhaycox** [comment]
Added tests for Rules
Added tests for Operations

Major linting exercise - sorry for all the changes

Fixed a minor issue with a missing space between the Caching^10% message
When creating a new value, the elementType is set to VARIABLE
Fixed SVG test warning
Fixed issue with 2 Reacts being loaded breaking hooks. Webpack configured to use external React and ReactDOM - same as ui-core
The linter raised a bunch of issue for unhandled Promises. In most cases I've just added a .catch((e) => console.error(e)), not ideal, but at least it won't fail silently now.

I ended up merging in Jose's state caching changes to ensure everything worked together and the appropriate linting was done on the merged code.

All tests run
Linting passes

Phew.

**Comment by ianhaycox** [comment]
I can create a PR for the initial code fixes no problem. Adding the tests might be tricky, but I'll give it a go and maybe have to remove/skip the broken tests.  Should I create a new Jira issue for the new PR ?

**Comment by jamesbrattBoomi** [comment]
Yeah, create a new Jira.

File: js/components/NoNetwork.tsx [hunk 5]
```diff
@@ -3,22 +3,16 @@ import { INoNetworkProps } from '../interfaces/INoNetwork';
 
 declare const manywho: any;
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/components/NoNetwork.tsx:4]: Done

File: js/components/Offline.tsx [hunk 6]
```diff
@@ -72,66 +62,77 @@ export class Offline extends React.Component<IOfflineProps, IOfflineState> {
...
-                    <div className="wait-spinner-small wait-spinner"></div>
-                    <span className="wait-message">Caching { String(this.props.cachingProgress) }%</span>
+            cachingSpinner = (
+                <div className="caching-spinner">
+                    <div className="wait-container">
+                        <div className="wait-spinner-small wait-spinner" />
+                        <span className="wait-message">
+                            { 'Caching ' }
+                            { String(this.props.cachingProgress) }
+                            %
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/components/Offline.tsx:95]: Done

File: js/components/Request.tsx [hunk 7]
```diff
@@ -78,40 +90,28 @@ class Request extends React.Component<IRequestProps, Partial<IRequestState>> {
...
-    }
+            .then((resp) => {
+
+                // This is to accomodate the scenario of
+                // modifying objectdata that had been created whilst offline
+                // that then needs to be saved to a service
+                extractExternalId(cachedRequest, tenantId, authenticationToken, stateId)
+                    .then(() => {
+                        this.onReplayResponse(resp);
+                    }, () => {});
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/components/Request.tsx:101]: OK

File: js/components/Offline.tsx [hunk 8]
```diff
@@ -72,66 +62,77 @@ export class Offline extends React.Component<IOfflineProps, IOfflineState> {
...
-                this.props.toggleIsOffline({ hasNetwork: false });
-                this.props.toggleIsReplaying(false);
-                this.setState({ view: null });
-            });
-    }
+            // Back into offline mode
+            this.props.toggleIsOffline({ hasNetwork: false });
+            this.props.toggleIsReplaying(false);
+            this.setState({ view: null });
+        });
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/components/Offline.tsx:77]: Added

File: js/components/GoOnline.tsx [hunk 9]
```diff
@@ -5,26 +5,23 @@ import { IGoOnlineProps, IGoOnlineState } from '../interfaces/IGoOnline';
...
+const mapStateToProps = (state: IGoOnlineState) => state;
 
-const mapDispatchToProps = (dispatch) => {
-    return {
-        toggleIsReplaying: (bool) => {
-            dispatch(isReplaying(bool));
-        },
-    };
-};
+const mapDispatchToProps = (dispatch) => ({
```
[the remark this answers is not in the document]
Comment (ianhaycox) [reply js/components/GoOnline.tsx:14]: Changed as discussed
